src/filerepository.js

```diff
@@ -0,0 +1,455 @@
+	 * occurs or if read process is aborted.
+	 */
+	upload() {
+		if ( this.status != 'idle' ) {
+			throw new CKEditorError( 'filerepository-upload-wrong-status: You cannot call upload if the status is different than idle.' );
+		}
+
+		this.status = 'uploading';
+
+		return this._adapter.upload()
+			.then( data => {
+				this.uploadResponse = data;
+				this.status = 'idle';
+
+				return data;
+			} )
+			.catch( err => {
+				if ( err === 'aborted' ) {
+					this.status = 'aborted';
+					throw 'aborted';
+				}
+
+				this.status = 'error';
+				throw err;
+			} );
+	}
+
+	/**
+	 * Aborts loading process.
+	 */
+	abort() {
+		if ( this.status == 'reading' ) {
+			this._reader.abort();
+		}
+
+		if ( this.status == 'uploading' && this._adapter.abort ) {
+			this._adapter.abort();
+		}
+
+		this.status = 'aborted';
+		this._destroy();
+	}
+
+	/**
+	 * Performs cleanup.
+	 *
+	 * @private
+	 */
+	_destroy() {
+		this._reader = undefined;
+		this._adapter = undefined;
+		this.data = undefined;
+		this.uploadResponse = undefined;
+		this.file = undefined;
+	}
+}
+
+mix( FileLoader, ObservableMixin );
+
+/**
+ * Adapter interface used by FileRepository to handle file upload. Adapter is a bridge between the editor and server that
+ * handles file uploads. It should contain logic necessary to initiate upload process and monitor its progress.
+ *
+ * It should implement two methods:
+ * * {@link module:upload/filerepository~Adapter#upload upload()},
+ * * {@link module:upload/filerepository~Adapter#abort abort()}.
+ *
+ * Example adapter implementation:
+ *
+ *	class Adapter {
+ *		constructor( loader ) {
+ *			// Save Loader instance to update upload progress.
+ *			this.loader = loader;
+ *		}
+ *
+ *		upload() {
+ *			// Update loader's progress.
+ *			server.onUploadProgress( data => {
+ *				loader.uploadTotal = data.total;
+ *				loader.uploaded = data.uploaded;
+ *			} ):
+ *
+ *			// Return promise that will be resolved when file is uploaded.
+ *			return server.upload( loader.file );
+ *		}
+ *
+ *		abort() {
+ *			// Reject promise returned from upload() method.
+ *			server.abortUpload();
+ *		}
+ *	}
+ *
+ * @interface Adapter
+ */
+
+/**
+ * Executes the upload process.
+ * This method should return a promise that will resolve when data will be uploaded to server. Promise should be
+ * resolved with an object containing information about uploaded file:
+ *
+ *	{
+ *		original: 'http://server/orginal-size.image.png'
+ *	}
+ *
+ * Take a look at {@link module:upload/filerepository~Adapter example Adapter implementation} and
+ * {@link module:upload/filerepository~FileRepository#createAdapter createAdapter method}.
+ *
+ * @method module:upload/filerepository~Adapter#upload
+ * @returns {Promise} Promise that should be resolved when data is uploaded.
+ */
+
+/**
+ * Aborts the upload process.
+ * After aborting it should reject promise returned from {@link #upload upload()} method with "aborted" string.
```

Comment:
There is no `return` in the sample. And I'm not sure if it is needed. If we assume that abort is always executed by the `loader.abort()` method, the status will change to `aborted` and the adapter developer should not need to do anything more.